### DISCUSSION TITLE
Some questions about Milvus 2:  1. CPU Usage 2. Recall Rate 3. Segment size

### DISCUSSION THREAD

**Discussion by owlwang** [milvus-io/milvus#18512] [Q&A and General discussion]
1. In version 2.1 the search seems to occupy the CPU does not seem to be utilized at all, only one or two core in the search, QPS is very low.
2. We tested the recall rate of milvus 1 and 2 search, the **recall rate is 20% lower** under the same conditions (FLAT without index)
3. How to control the size of each segment in version 2.1?
     We have modified dataCoord.segment.maxSize and dataCoord.segment.segment.sealProportion, but we still can't control the number and size of segments built.
4. Will milvus provide an http restful api interface in the future?

Below is the Chinese translation (以下为中文)：

我们在从 milvus 1 转向到 2 遇到了几个问题，想请教下您们

1. 在2.1版本中搜索的时候似乎占用 CPU 似乎完全没有利用起来，只占了一两个核心在搜索，QPS 非常低。
2. 我们测试了 milvus 1 和 2 的搜索的召回率，相同条件下召回率低了20% （FLAT无索引）
3. 在2.1版本中如何控制每个 segment 的大小？
     我们修改了dataCoord.segment.maxSize和 dataCoord.segment.sealProportion 似乎还是没法控制建库的segment 的数量和大小。
4. Milvus 在未来会提供 http restful 接口吗？

**Comment by yhmo** [comment]
1. We are preparing a benchmark for v2.1, could you provide more details of your QPS testing(how many rows, dimensions, what kind of index, index/search parameters, the client test script, etc..)?
2. FLAT is brute-force search, the result should be same, could you provide more details of the testing?
3. dataCoord.segment.maxSize is the major parameter to control segment size, by default the value is 512MB. Milvus has a compaction machinery to merge small segments into large segment, sometimes a large segment size may exceed this value. Segment size cannot be controlled precisely.
4. The v2.1 already support http restful api, in the milvus.yml the **proxy.http.enabled** is for this function:
```
proxy:
  port: 19530
  internalPort: 19529
  http:
    enabled: true # Whether to enable the http server
```
But the documentation is not ready, we have written a guideline for it, it will be published recently.

**Comment by owlwang** [comment]
@yhmo 
1. we have a total of about 13 million 128-dimensional vectors, index sq8, nprobe 128, test script written in python. (This we later found in the multi-threaded test qps utilization rate has increased, may be the problem of the method we use, we are doing further testing, will post the details)
2. we found that the recall rate was 20% lower after testing sq8, in order to exclude the error caused by the index, we re-tested with flat and it was still 20% lower
3. we found a serious problem, milvus does not seem to create indexes when we set the max size of segment to 16384 (i.e. 16G) (we are inserting first and then building order). The index is created in seconds, but it doesn't seem to be created, and an error is reported in the log when this index is deleted.
4. By the way, is there an interface to see how many segments there are now?

_[2022/08/03 09:27:24.384 +00:00] [ERROR] [proxy/task_scheduler.go:468] ["Failed to execute task: index not exist"] [traceID=7282872868e0ed6a] [stack="github.com/milvus-io/milvus/internal/proxy.(*taskScheduler).processTask\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:468\ngithub.com/milvus-io/milvus/internal/proxy.(*taskScheduler).definitionLoop\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:494"]_


To summarize the parameters and details used in our third question, the others are measured in the default parameters:

- 13 million vectors
- index ivf-sq8
- nprobe 128
- segment_max_size = 16384(16G)
- nlist 8192


Below is the Chinese translation (以下为中文)：


1. 我们总共大约有1300万个128维向量，index sq8, nprobe 128，测试脚本拿 python 写的。（这个我们后来在多线程测试中发现 qps 利用率有上升，可能是我们使用方法的问题，我们在做进一步的测试，会将详细信息贴过来)
2. 我们在测试过 sq8 之后发现召回率低了20%，为了排除索引带来的误差，用 flat 重新测得还是低20%
3. 我们发现了一个严重问题, 在我们将segment max size 设置为16384（即 16G）时，milvus 似乎不会创建索引（我们是先插再建的顺序）。创建索引秒创建，但是实际上似乎没有创建，删除这个索引的时候日志中报了一个 error 的错误。
4. 顺便问一下有接口可以看到现在有多少个 segment 吗？

_[2022/08/03 09:27:24.384 +00:00] [ERROR] [proxy/task_scheduler.go:468] ["Failed to execute task: index not exist"] [traceID=7282872868e0ed6a] [stack="github.com/milvus-io/milvus/internal/proxy.(*taskScheduler).processTask\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:468\ngithub.com/milvus-io/milvus/internal/proxy.(*taskScheduler).definitionLoop\n\t/go/src/github.com/milvus-io/milvus/internal/proxy/task_scheduler.go:494"]_

总结一下我们第三个问题中使用的参数和详情, 其他的都是在默认参数测的：

- 1300万条向量
- 索引 ivf-sq8
- nprobe 128
- segment_max_size = 16384(16G)
- nlist 8192

**Reply by yhmo** [reply]
Our test team will verify these problems.
In pymilvus, the api **utility.get_query_segment_info()** can get all segments information that loaded in query node memory.
```
def get_query_segment_info(collection_name, timeout=None, using="default"):
    """
    Notifies Proxy to return segments information from query nodes.

    :param collection_name: A string representing the collection to get segments info.
    :param timeout: An optional duration of time in seconds to allow for the RPC. When timeout
                    is set to None, client waits until server response or error occur.
    :type  timeout: float

    :return: QuerySegmentInfo:
        QuerySegmentInfo is the growing segments's information in query cluster.
    :rtype: QuerySegmentInfo

    :example:
        >>> from pymilvus import Collection, FieldSchema, CollectionSchema, DataType, connections, utility
        >>> connections.connect(alias="default")
        >>> _DIM = 128
        >>> field_int64 = FieldSchema("int64", DataType.INT64, description="int64", is_primary=True)
        >>> field_float_vector = FieldSchema("float_vector", DataType.FLOAT_VECTOR, description="float_vector", is_primary=False, dim=_DIM)
        >>> schema = CollectionSchema(fields=[field_int64, field_float_vector], description="get collection entities num")
        >>> collection = Collection(name="test_get_segment_info", schema=schema)
        >>> import pandas as pd
        >>> int64_series = pd.Series(data=list(range(10, 20)), index=list(range(10)))i
        >>> float_vector_series = [[random.random() for _ in range _DIM] for _ in range (10)]
        >>> data = pd.DataFrame({"int64" : int64_series, "float_vector": float_vector_series})
        >>> collection.insert(data)
        >>> collection.load() # load collection to memory
        >>> res = utility.get_query_segment_info("test_get_segment_info")
    """
```

We have add guideline for restful api in the official documentation: https://milvus.io/docs/v2.1.x/create_collection.md , open this page you will see each code block has a "RESTful API" panel.

**Comment by owlwang** [comment]
We re-tested the recall drop mentioned in the second question, we used the wrong model, we are sorry.

我们重新测了下 第二个问题提到的召回率下降的问题，经过确认是我们用错了模型了，抱歉。

**Comment by binbinlv** [comment]
For question 3, actually, as your config, it is not reproduced in the latest master build: 
milvus: [master-20220805-c267ad6](https://hub.docker.com/layers/milvus/milvusdb/milvus/master-20220805-c267ad6/images/sha256-771c018333567688b6d751c3de9836a87c37cd9c83249a4a4f5a8262e0056204?context=explore)
pymilvus: 2.1.0

dataCoord.segment.maxSize = 16384
```
 segment:
    maxSize: "16384"  # Maximum size of a segment in MB
    sealProportion: 0.75  # It's the minimum proportion for a segment which can be sealed
    assignmentExpiration: 2000  # ms
```
dim = 128
insert data: 13 million
index: ivf-sq8, nlist 8192

And at my side, the interface "create_index" needs some time to return, not return in seconds.

@owlwang what is the specific version of the milvus and pymilvus you are using? And could you have a look whether there is any difference between your script and mine?

The script is:
```
>>> from pymilvus import CollectionSchema, FieldSchema
>>> from pymilvus import Collection
>>> from pymilvus import connections
>>> from pymilvus import DataType
>>> from pymilvus import Partition
>>> from pymilvus import utility
>>>
>>> dim = 128
>>> connections.connect(host="***", port="19530")
>>> collection = Collection("test_search_collection_binbin_1")
>>> nb =1000000
>>> import random
>>> vectors = [[random.random() for _ in range(dim)] for _ in range(nb)]
>>> import numpy as np
>>> for i in range(0, 13):
...           res = collection.insert([[i for i in range(i*nb, (i+1)*nb)], [np.float32(i) for i in range(i*nb, (i+1)*nb)], vectors])
...
>>>
>>> index_param = {"index_type": "IVF_SQ8", "metric_type": "L2", "params": {"nlist": 8192}}
>>> collection.create_index("float_vector", index_param)

Status(code=0, message='')
>>>
>>>
>>> collection.has_index()
True
>>> collection.indexes
[<pymilvus.orm.index.Index object at 0x7f85c3a9c310>]
>>> collection.index()
<pymilvus.orm.index.Index object at 0x7f85c3a9aa60>
>>>
>>>
>>> collection.drop_index()
>>>
>>> collection.has_index()
False
```

**Reply by owlwang** [reply]
sealProportion=0.25

The index I created in Antu.


```python3
fields = [
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=DIM)
]

class InsertQ:
    def __init__(self):
        self.ids_buf = []
        self.vecs_buf = []
        self.buffer_size = 5000

    def add(self, m_id, m_vec):
        if m_id and m_vec:
            self.ids_buf.append(m_id)
            self.vecs_buf.append(m_vec)
        if len(self.ids_buf) >= self.buffer_size:  # buffer size
            self.flush()

    def flush(self):
        print(f'committing to db {len(self.ids_buf)}')

        entities = [
            self.ids_buf,
            self.vecs_buf,  # field random, only supports list
        ]

        result = milvus_partition.insert(entities)
        print(result)

        self.ids_buf = []
        self.vecs_buf = []
        print(f'finished commit')

    def cleanup(self):
        if len(self.ids_buf):
            self.flush()
        print(f'cleanup')

```

**Reply by binbinlv** [reply]
> sealProportion=0.25
> 
> The index I created in Antu.
> 
> ```python
> fields = [
>     FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
>     FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=DIM)
> ]
> 
> class InsertQ:
>     def __init__(self):
>         self.ids_buf = []
>         self.vecs_buf = []
>         self.buffer_size = 5000
> 
>     def add(self, m_id, m_vec):
>         if m_id and m_vec:
>             self.ids_buf.append(m_id)
>             self.vecs_buf.append(m_vec)
>         if len(self.ids_buf) >= self.buffer_size:  # buffer size
>             self.flush()
> 
>     def flush(self):
>         print(f'committing to db {len(self.ids_buf)}')
> 
>         entities = [
>             self.ids_buf,
>             self.vecs_buf,  # field random, only supports list
>         ]
> 
>         result = milvus_partition.insert(entities)
>         print(result)
> 
>         self.ids_buf = []
>         self.vecs_buf = []
>         print(f'finished commit')
> 
>     def cleanup(self):
>         if len(self.ids_buf):
>             self.flush()
>         print(f'cleanup')
> ```

so here "milvus_partition.insert(entities)" means you use partition.insert(), not collection.insert(), right?

**Reply by owlwang** [reply]
yes

**Reply by owlwang** [reply]
milvus_partition = milvus_collection.create_partition(partition_name, f"{partition_name} data")

**Reply by owlwang** [reply]
remind: The index I created in Antu.

**Reply by binbinlv** [reply]
I tested it again using milvus v2.1.0 of my script, it is good too.

So you totally have two partitions (default and one partition you created)? And insert data in the partition you created.

Then using collection.create_index to create index? 

I will try to recreate this issue like this operations.

**Reply by owlwang** [reply]
I re-inserted it into the default partition using collection.insert(), still have the problem.

Also I record a video

https://user-images.githubusercontent.com/13383393/183066929-c9c61b33-44ea-42e6-a8a0-e99e6e66c42c.mp4

**Reply by binbinlv** [reply]
From the video, it seems just could see you created the index, but I did not see any errors?

You mean you create it and drop it immediately? Then it reports error "the index does not exist"?

**Reply by owlwang** [reply]
This second creation does not make sense, I see the words "index does not exist" in the logs, I am not sure if the index does not really exist. But normally it would show a progress bar instead of what it does now.

Then when I load, the Antu interface doesn't work at all, it can't read the database information, it just reports "syncTimestamp Failed:context canceled". I think there is something wrong with this.

**Reply by binbinlv** [reply]
So may I know what is your milvus version? 2.1.0? And standalone or cluster? I will try with attu.

**Reply by owlwang** [reply]
milvus 2.1 standalone

**Comment by yhmo** [comment]
@owlwang 
Just remind, we have provided guideline for RESTful API. Open this page: https://milvus.io/docs/v2.1.x/create_collection.md, there is a "RESTful API" panel in each code block.

**Reply by owlwang** [reply]
Thanks 🍻

**Comment by soothing-rain** [comment]
Hey @owlwang , segment size == 16GigBytes might be a little too much if you don't have a lot of resources. May I ask why you choose to set the segment size to 16GigBytes? Creating indexed on a huge segment might require a lot of memory resources, and would fail if there's not enough resource.

Can you provide the more of your testing script? For example when was `create_index` method called? Is it call before or after your insert?

**Reply by owlwang** [reply]
The reason is that we have a billion of data, and we have calculated that this value is reasonable in standalone mode.(This is how we use it in milvus 1.1)

we call create_index after insert.

Here is a pseudo code that we inserted, We will use Antu to create the index after we have finished inserting it.

```
import sys
import mmap
from pymilvus import (
    connections,
    utility,
    FieldSchema, CollectionSchema, DataType,
    Collection,
)

HOST = "127.0.0.1"
PORT = "19530"
DIM = 128

collection_name = 'test'
partition_name = 'test_partition'

print("start connecting to Milvus")
connections.connect("default", host=HOST, port=PORT)

has_collection = utility.has_collection(collection_name)
print(f"Does collection {collection_name} exist in Milvus: {has_collection}")

fields = [
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=DIM)
]

schema = CollectionSchema(fields, "Milvus")

milvus_collection = Collection(collection_name, schema, consistency_level="Strong")

has_partition = utility.has_partition(collection_name, partition_name)

print(f"Does collection {partition_name} exist in collection_name {collection_name}: {has_partition}")

milvus_partition = milvus_collection.create_partition(partition_name, f"{partition_name} data")

class InsertQ:
    def __init__(self):
        self.ids_buf = []
        self.vecs_buf = []
        self.buffer_size = 100000

    def add(self, m_id, m_vec):
        if m_id and m_vec:
            self.ids_buf.append(m_id)
            self.vecs_buf.append(m_vec)
        if len(self.ids_buf) >= self.buffer_size:  # buffer size
            self.flush()

    def flush(self):
        print(f'committing to db {len(self.ids_buf)}')

        entities = [
            self.ids_buf,
            self.vecs_buf,  # field random, only supports list
        ]

        result = milvus_partition.insert(entities)
        print(result)
        self.ids_buf = []
        self.vecs_buf = []
        print(f'finished commit')

    def cleanup(self):
        if len(self.ids_buf):
            self.flush()
        print(f'cleanup')


insert_q = InsertQ()


def read_feat(fn):
    # read features
    return [[]]


def gen_id(fn, index):
    # calc id
    return 123456


def insert_one_file(filename):
    # one file has multiple vectors
    feats = read_feat(filename)
    index = 0
    for feat in feats:
        m_id = gen_id(filename, index)
        insert_q.add(m_id, feat)
        index += 1


def insert_lst(lst_filename):
    with open(lst_filename, 'r') as r:
        for line in r:
            np_filename = line.strip()
            insert_one_file(np_filename)

    insert_q.cleanup()


if __name__ == '__main__':
    insert_lst(sys.argv[1])

````

**Reply by soothing-rain** [reply]
Thanks for your reply.

1) About the segment size, what you do mean by "reasonable"? Are you setting it to 16Gigs to speed up searching?

2) About index building, Attu calls the asynchronous interface to build indexes. Attu returns immediately unless an error occurs. Then index building will be going on in the background, waiting for resources to build index (in your case, I think Milvus will just wait forever as there will be no new index node coming in, nor increased memory).

**Reply by owlwang** [reply]
1. yes, for speed up.
2. Got it

**Reply by soothing-rain** [reply]
I see. But since you have 1300million 128-d vectors which adds up to ~6.2GigBytes. Why dont we start with small segment size and go from there? :)

**Reply by owlwang** [reply]
This is just a copy of the parameters from our milvus 1.0 which has **a billion** vectors. Of course, putting them all in one segment may have an impact.

We also tested the others (4096 and the default 512), but as I mentioned in #18530, we were unable to complete this test.

And our main concern is the performance of the query.
